pkg/events/options.go

```diff
@@ -0,0 +1,95 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"net"
+	"sort"
+	"strings"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/auth"
+	"go.thethings.network/lorawan-stack/v3/pkg/rpcmetadata"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"google.golang.org/grpc/peer"
+)
+
+// Option is an option that is used to build events.
+type Option interface {
+	applyTo(*event)
+}
+
+type optionFunc func(e *event)
+
+func (f optionFunc) applyTo(e *event) { f(e) }
+
+// WithIdentifiers returns an option that sets the identifiers of the event.
+func WithIdentifiers(identifiers ...ttnpb.Identifiers) Option {
+	return optionFunc(func(e *event) {
+		for _, ids := range identifiers {
+			e.innerEvent.Identifiers = append(e.innerEvent.Identifiers, ids.EntityIdentifiers())
+		}
+	})
+}
+
+// WithData returns an option that sets the data of the event.
+func WithData(data interface{}) Option {
+	return optionFunc(func(e *event) {
+		e.data = data
```

Comment:
this should panic if `e.data` is not nil

Reply:
I'm just moving what was already there. The goal here is not to reimplement.

Reply:
That is not true - before `data` could only be specified once, now it's possible to specify it multiple times - you changed the semantics